Implement iif function

johngrimes opened the pull request:
Resolves #195.

piotrszul commented on fhir-server/src/main/java/au/csiro/pathling/fhirpath/function/IfFunction.java:
```diff
@@ -0,0 +1,90 @@
+            ifTrueType);
+  }
+
+  private FHIRDefinedType getFhirTypeForArgument(final FhirPath argument) {
```
@Nonnull missin

johngrimes replied:
Fixed.

piotrszul commented on fhir-server/src/main/java/au/csiro/pathling/fhirpath/literal/CodingLiteralPath.java:
```diff
@@ -50,29 +51,54 @@ protected CodingLiteralPath(@Nonnull final Dataset<Row> dataset, @Nonnull final
+
+  @Nonnull
+  private static String encodeComponent(@Nonnull final String component) {
+    if (component.matches("[',]")) {
```
This looks suspicious: 
Why ',' should indicate need for escaping.
Also probably should be component.matches(".*[',].*")

johngrimes replied:
I've changed this to match the same pattern that is used for `RAW_CODING_COMPONENT`, which is any character except those used to delimit strings, arguments and coding components.

piotrszul commented on fhir-server/src/main/java/au/csiro/pathling/fhirpath/literal/CodingLiteralPath.java:
```diff
@@ -50,29 +51,54 @@ protected CodingLiteralPath(@Nonnull final Dataset<Row> dataset, @Nonnull final
+  @Nonnull
+  private static String decodeComponent(@Nonnull final String component) {
+    if (component.matches("^'(.*)'$")) {
+      final String result = unSingleQuote(component);
```
The regex can be probably precompiled for better perfomance.

johngrimes replied:
Done.

piotrszul commented on fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/IfFunctionTest.java:
```diff
@@ -0,0 +1,384 @@
+        .withRow("observation-4", makeEid(1), "bar")
+        .withRow("observation-5", makeEid(0), "bar")
+        .withRow("observation-5", makeEid(1), "bar")
+        .build();
```
I can see that we treat null values in the condition as 'false' but is that correct? 
Should we not rather preserve nulls?

johngrimes replied:
I think that is the case for most FHIRPath functions, but the spec for this one says:

> If criterion is false or an empty collection, the function returns otherwise-result, unless the optional otherwise-result is not given, in which case the function returns an empty collection.

johngrimes replied:
Note that we don't do the "optional" part currrently - I'm not really sure what the value of it is.

Doesn't leaving out the third argument just turn this into `where`?

piotrszul reviewed (commented):
Looks good to me except for a few comments inside the code.

johngrimes commented:
@piotrszul I found some cases that weren't covered by my original implementation - things like having `ResourcePath` or `UntypedResourcePath` in the arguments, or `BackboneElements`.

I have fixed this by introducing a method called FhirPath.mergeWith which knows how to create a new path which can be used to accommodate values from both the invoking path and a target path.

I've also added a bunch more test coverage.

piotrszul commented on fhir-server/src/main/java/au/csiro/pathling/fhirpath/FhirPath.java:
```diff
@@ -47,4 +48,25 @@
+  NonLiteralPath mergeWith(FhirPath target, Dataset<Row> dataset, String expression,
+      Column idColumn, Optional<Column> eidColumn, Column valueColumn, boolean singular,
+      Optional<Column> thisColumn);
+
```
Add Nonnull annotations?

johngrimes replied:
I can't seem to be able to get IntelliJ to tell me about this sort of issue.

Maybe it's because this is an interface...

johngrimes replied:
Actually, maybe this would have picked it up (in a kind of indirect way):

<img width="422" alt="image" src="https://user-images.githubusercontent.com/55939/106849632-c2271680-66fe-11eb-86a5-0708e63eebeb.png">

johngrimes replied:
Fixed in c6592b2.

piotrszul commented on fhir-server/src/main/java/au/csiro/pathling/fhirpath/UntypedResourcePath.java:
```diff
@@ -123,4 +124,23 @@ public UntypedResourcePath copy(@Nonnull final String expression,
+    if (target instanceof UntypedResourcePath && possibleTypes
+        .equals(((UntypedResourcePath) target).possibleTypes)) {
+      // Two UntypedResourcePaths can be merged together if they have the same set of possible
+      // types.
```
I think this could potentially be more lenient, that it the any combination of possibleTypes should be allowed by the results possibleTypes should be the union of both inputs.
Or per minimum it what should be allowed is merging when possibleTypes contain target.possibleTypes.

johngrimes replied:
Interesting... as far as I can tell, we never actually use `possibleTypes`. And I think this is actually ok, any resources which don't match just don't get added to the resulting collection.

So you can pass anything you like as an argument to `ofType` - as long as it is a valid resource type code. It is not actually checked against the `possibleTypes` of the input path.

The spec doesn't really say that you need to check, from my reading: http://hl7.org/fhirpath/#oftypetype-type-specifier-collection

So I actually think we could remove `possibleTypes` altogether... unless I am missing something.

johngrimes replied:
I've removed `possibleTypes` in e99e70a.

piotrszul commented on fhir-server/src/test/java/au/csiro/pathling/fhirpath/literal/CodingLiteralPathTest.java:
```diff
@@ -0,0 +1,135 @@
+
+    final String actualExpression = codingLiteralPath.getExpression();
+    assertEquals(expression, actualExpression);
+  }
```
But is this correct? 
If code value "46,2" should be quoted and escaped, should not he actualExpression have be:
" ... |`46\\,2"? 
Also to test escaping should the expression be " ... |`46\\,2" as well?

johngrimes replied:
I don't think commas need to be escaped in the context of a string literal: http://hl7.org/fhirpath/#string. They are, however, problematic _outside_ of a string literal.

If we had a Coding with a code that contained a comma that we wanted to encode as a Coding literal, we would need to quote the code to ensure that the parser didn't interpret it as the end of the Coding literal (and maybe the start of a new argument, if it was in the context of a function's arguments). The same goes for whitespace, parentheses and the pipe operator.

So `NEEDS_QUOTING` needs to be the minimal set of characters that indicate that the component will need to be quoted so that it doesn't fall foul of the patterns in `FhirPath.g4`.

piotrszul replied:
Ok, I guess I did not fully understand what needs to be quoted/escapes.
But do we have a test case for such a coding the needs escaping?

johngrimes replied:
I have added more tests in 8ea952a.

piotrszul reviewed (commented):
Looks good apart from a few comments in the code.
